### PR TITLE
Fix compatibility with Python 3.7

### PR DESCRIPTION
**Pull request by felixonmars**

### CODE REVIEW

**Comment by rbarrois** [comment]
Hey, welcome to the project!

That's an interesting fix, do you have any pointers to the Python Changelog — why would this change be needed?

And we might have more issues to fix; we should run the test suite against Python 3.7 — a good inspiration could be commit e55493d7aa7aebead8fd5f689503477cb0c00e0e, which added support for pypy3.
Would you mind helping on that topic?

Thanks!

**Comment by felixonmars** [comment]
The issue is an intentional regression introduced in Python 3.7 as part of PEP 479: https://www.python.org/dev/peps/pep-0479/

I'll update the tests.

**Comment by felixonmars** [comment]
Looks like Python 3.7 is not yet supported in Travis yet. Do you want to use nightly for the tests?

**Comment by rbarrois** [comment]
That would be awesome, yes :)

Thanks!

**Comment by felixonmars** [comment]
I guess that's how it works now :)

File: factory/utils.py [hunk 1]
```diff
@@ -133,7 +133,10 @@ def __iter__(self):
             if self.next_elements:
                 yield self.next_elements.popleft()
             else:
-                value = next(self.iterator)
+                try:
+                    value = next(self.iterator)
+                except StopIteration:
```
Comment (kingbuzzman) [inline factory/utils.py]: Indeed it does crash on 3.7, this fixes it. One thing id like to see is adding 3.7 to tox so this kind of thing is caught sooner. Current error on master:

```
======================================================================
ERROR: test_no_reset (tests.test_utils.ResetableIteratorTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/factory_boy/factory/utils.py", line 136, in __iter__
    value = next(self.iterator)
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/factory_boy/tests/test_utils.py", line 269, in test_no_reset
    self.assertEqual([1, 2, 3], list(i))
RuntimeError: generator raised StopIteration

======================================================================
ERROR: test_reset_after_end (tests.test_utils.ResetableIteratorTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/factory_boy/factory/utils.py", line 136, in __iter__
    value = next(self.iterator)
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/factory_boy/tests/test_utils.py", line 324, in test_reset_after_end
    self.assertRaises(StopIteration, next, iterator)
  File "/usr/local/lib/python3.7/unittest/case.py", line 743, in assertRaises
    return context.handle('assertRaises', args, kwargs)
  File "/usr/local/lib/python3.7/unittest/case.py", line 178, in handle
    callable_obj(*args, **kwargs)
RuntimeError: generator raised StopIteration
```

How to test it:
```
cat << EOF | docker run --rm -i python:3.7 bash
git clone https://github.com/FactoryBoy/factory_boy.git
cd factory_boy/
pip install -r requirements_test.txt 
pip install django==1.11 SQLAlchemy==1.2 mongoengine==0.15 flask flask_sqlalchemy
pip install -e .
apt-get update
yes | apt-get install mongodb
mkdir -p /data/db
mongod > /dev/null &
make test
EOF
```

**Comment by francoisfreitag** [comment]
Python 3.7 support for Factory Boy is exciting, thank you for the PR!

I rebased your work on master and tried to make Travis CI run the tests on Python 3.7. I believe there's only a small edit to make on `.travis.yml`:

```diff
diff --git a/.travis.yml b/.travis.yml
index 6c574b1..9553f42 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -18,8 +18,10 @@ matrix:
       env: TOXENV=py35-django20-alchemy12-mongoengine015
     - python: "3.6"
       env: TOXENV=py36-django20-alchemy12-mongoengine015
-    - python: "nightly"
+    - python: "3.7"
       env: TOXENV=py37-django20-alchemy12-mongoengine015
+      dist: xenial
+      sudo: true
 
     # Django LTS
     - python: "3.6"
```

This is taken from https://github.com/travis-ci/travis-ci/issues/9815#issuecomment-401756442.
Here's my [test PR](https://github.com/francoisfreitag/factory_boy/pull/1) and the [corresponding travis build](https://travis-ci.org/francoisfreitag/factory_boy/builds/417971929).

---
I believe the commit messages could reference https://github.com/FactoryBoy/factory_boy/issues/492. We should be able to close that issue when this pull request is merged.

**Comment by francoisfreitag** [comment]
In addition to the above diff, it may be a good idea to reference the travis issue https://github.com/travis-ci/travis-ci/issues/9815 in a comment in `.travis.yml`, so that we occasionally verify that this workaround is still required.

```diff
diff --git a/.travis.yml b/.travis.yml
index 6c574b1..92d57ff 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -18,8 +18,11 @@ matrix:
       env: TOXENV=py35-django20-alchemy12-mongoengine015
     - python: "3.6"
       env: TOXENV=py36-django20-alchemy12-mongoengine015
-    - python: "nightly"
+    - python: "3.7"
       env: TOXENV=py37-django20-alchemy12-mongoengine015
+      # Workaround https://github.com/travis-ci/travis-ci/issues/9815.
+      dist: xenial
+      sudo: true
 
     # Django LTS
     - python: "3.6"
```

**Comment by jeffwidman** [comment]
Thanks for this. I ended up merging #511 instead because it incorporated both the PR and the patches suggested by reviewers. Sorry you don't get the credit on the commit history @felixonmars but in my heart you're a winner and we both know you were the source of this. I thought about holding off merging, but decided the other one was ready and I hadn't seen any activity from @felixonmars since July, so errored on shipping.

**Comment by felixonmars** [comment]
Oops, somehow missed the update :P

**Pull request closed without merging**